sdks/java/io/neo4j/src/main/java/org/apache/beam/sdk/io/neo4j/Neo4jIO.java

```diff
@@ -0,0 +1,1221 @@
+        @NonNull String cypher,
+        @Nullable SerializableFunction<ParameterT, Map<String, Object>> parametersFunction,
+        long batchSize,
+        boolean logCypher,
+        String unwindMapName) {
+      super(driverProviderFn, sessionConfig, transactionConfig);
+      this.cypher = cypher;
+      this.parametersFunction = parametersFunction;
+      this.logCypher = logCypher;
+      this.batchSize = batchSize;
+      this.unwindMapName = unwindMapName;
+
+      unwindList = new ArrayList<>();
+
+      elementsInput = 0;
+      loggingDone = false;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      // Map the input data to the parameters map...
+      //
+      ParameterT parameters = context.element();
+      if (parametersFunction != null) {
+        // Every input element creates a new Map<String,Object> entry in unwindList
+        //
+        unwindList.add(parametersFunction.apply(parameters));
+      } else {
+        // Someone is writing a bunch of static or procedurally generated values to Neo4j
+        unwindList.add(Collections.emptyMap());
+      }
+      elementsInput++;
+
+      if (elementsInput >= batchSize) {
+        // Execute the cypher query with the collected parameters map
+        //
+        executeCypherUnwindStatement();
+      }
+    }
+
+    private void executeCypherUnwindStatement() {
+      // In case of errors and no actual input read (error in mapper) we don't have input
+      // So we don't want to execute any cypher in this case.  There's no need to generate even more
+      // errors
+      //
+      if (elementsInput == 0) {
+        return;
+      }
+
+      // Add the accumulated list to the overall parameters map
+      // It contains a single parameter to unwind
+      //
+      final Map<String, Object> parametersMap = new HashMap<>();
+      parametersMap.put(unwindMapName, unwindList);
+
+      // Every "write" transaction writes a batch of elements to Neo4j.
+      // The changes to the database are automatically committed.
+      //
+      TransactionWork<Void> transactionWork =
+          transaction -> {
+            Result result = transaction.run(cypher, parametersMap);
+            while (result.hasNext()) {
+              // This just consumes any output but the function basically has no output
+              // To be revisited based on requirements.
+              //
+              result.next();
+            }
+            return null;
+          };
+
+      if (logCypher && !loggingDone) {
+        String parametersString = getParametersString(parametersMap);
+        LOG.info(
+            "Starting a write transaction for unwind statement cypher: "
+                + cypher
+                + ", parameters: "
+                + parametersString);
+        loggingDone = true;
+      }
+
+      if (driverSession.session == null) {
+        throw new RuntimeException("neo4j session was not initialized correctly");
```

Comment:
Are there cases where we can just retry instead of throwing an exception here ? Throwing  here usually means a runner-initiated retry which should should be correct could be limited (for example fail the job after four retries) and might not use exponential backoff.

Reply:
s / which should should be correct could be limited / which should be correct but could be limited